jsonpickle/ext/numpy.py

```diff
@@ -58,18 +65,175 @@ class NumpyNDArrayHandler(NumpyBaseHandler):
+
+    Furthermore, ndarrays which are views of array-like objects implementing __array_interface__,
+    but which are not themselves nd-arrays, are deepcopied with a warning,
+    as we cannot guarantee whatever custom logic such classes implement is correctly reproduced.
+    """
+    mode = 'warn'
+
+    def flatten(self, obj, data):
+        """encode numpy to json"""
+        base = obj.base
+        if base is None:
+            # store by value
+            data = super(NumpyNDArrayHandlerView, self).flatten(obj, data)
+        elif isinstance(base, np.ndarray) and base.data.contiguous:
+            # store by reference
+            self.flatten_dtype(obj.dtype, data)
+            data['base'] = self.context.flatten(base, reset=False)
+            # for a view, always store shape
+            data['shape'] = obj.shape
+
+            offset = obj.ctypes.data - base.ctypes.data
+            if offset:
+                data['offset'] = offset
+
+            if not obj.data.c_contiguous:
+                data['strides'] = obj.strides
+        else:
+            # store a deepcopy or fail
+            if self.mode == 'warn':
+                msg = "ndarray is defined by reference to an object we do not know how to serialize. " \
+                      "A deep copy is serialized instead, breaking memory aliasing."
+                warnings.warn(msg)
+            elif self.mode == 'raise':
+                msg = "ndarray is defined by reference to an object we do not know how to serialize."
+                raise ValueError(msg)
+            data = super(NumpyNDArrayHandlerView, self).flatten(obj.copy(), data)
+
+        if obj.flags.writeable is False:
+            data['writeable'] = False
+
+        return data
+
+    def restore(self, data):
+        """decode numpy from json"""
+        base = data.get('base', None)
+        if base is None:
+            # decode array with owndata=True
+            arr = super(NumpyNDArrayHandlerView, self).restore(data)
+        else:
+            # decode array view, which references the data of another array
+            base = self.context.restore(base, reset=False)
+            assert base.data.contiguous, \
+                "Current implementation assumes base is contiguous"
+
+            arr = np.ndarray(
+                buffer=base.data,
+                dtype=self.restore_dtype(data),
+                shape=data.get('shape'),
+                offset=data.get('offset', 0),
+                strides=data.get('strides', None)
+            )
+            if not data.get('writeable', True):
+                arr.flags.writeable = False
+
+
+        return arr
 
 
 def register_handlers():
     jsonpickle.handlers.register(np.dtype, NumpyDTypeHandler, base=True)
     jsonpickle.handlers.register(np.generic, NumpyGenericHandler, base=True)
-    jsonpickle.handlers.register(np.ndarray, NumpyNDArrayHandler, base=True)
+    jsonpickle.handlers.register(np.ndarray, NumpyNDArrayHandlerView, base=True)
```

Comment:
Just curious, does this break backwards-compatibility with an existing json file that serialized with the older handler? If so we'll probably want to add a compatibility warning and bump the minor version up on next release.  Hopefully this handler can take the old data too?